## Intro

Here I parse the lat-long coords of the stations so that we can visually examine stations on a map as well as generate some features about their locations such as zoning district as well as community population and demographics. Additionally I scrape weather data for 2015 to see how the weather affects ridership.

In [1]:
import zipfile
from subprocess import call

import psutil

import pandas as pd
import numpy as np
import json
import datetime
import re

import os
import cPickle as pickle
from multiprocessing import Pool

import geopy
import geopandas as gpd
import shapely.geometry as geom

import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
with open('sample4Months.pkl') as f:
    samp = pickle.load(f)
    
samp.shape

(3184288, 15)

### Aquiring Addresses

First, I'm going to get official addresses for our stations using geopy as well as convert our lat long coords into shapely geometry points.

In [3]:
# unique lat-long coords
lat_long = samp[~samp[['start station latitude', 'start station longitude']].duplicated()][['start station id','start station name','start station latitude', 'start station longitude']]

from geopy.geocoders import Nominatim, GoogleV3

geo = Nominatim()
goog = GoogleV3()
key = "AIzaSyB_aNmk-hv198hnDVSAxoLJocNo3fLw1-4"


# Convert lat-long to geopy addresses
# save progress in this dictionary
station_locs = {}

In [ ]:
# geopy needs a string
lat_long['lat_long'] = lat_long['start station latitude'].map(str) +', '+ lat_long['start station longitude'].map(str)

for idx in lat_long.index:
    if lat_long.ix[idx, 'start station id'] in station_locs:
        continue
    station_locs[lat_long.ix[idx, 'start station id']] = geo.reverse(lat_long.ix[idx, 'lat_long']).address

# save for easier loading
with open('station_locs.pkl', 'wb') as f:
    pickle.dump(station_locs, f, 2)

In [4]:
with open('station_locs.pkl', 'rb') as f:
    station_locs = pickle.load(f)

len(station_locs)

459

In [5]:
# Move results back to samp data frame

samp['lat_long'] = zip(samp['start station latitude'].values, samp['start station longitude'].values)
# Geopandas for some reason uses (longitude,latitude)
samp['long_lat'] = zip(samp['start station longitude'].values, samp['start station latitude'].values)
samp['address'] = samp['start station id'].map(station_locs)
samp['zipcode'] = samp.address.str.findall(r'\d{5}')
samp.loc[~samp.zipcode.isnull(), 'zipcode'] = samp.loc[~samp.zipcode.isnull(), 'zipcode'].map(lambda t:t[0] if (len(t)>0) else np.nan)


Some of the addresses and zip codes are null so lets try and fix that

In [6]:
samp.ix[samp.zipcode.isnull(), 'start station name'].unique()

array(['MacDougal St & Washington Sq', '48 Ave & 5 St',
       'Marcy Ave & MacDonough St', 'Central Park West & W 85 St',
       'W 76 St & Columbus Ave', '5 Ave & E 78 St',
       'East End Ave & E 86 St', 'Huron St & Franklin St',
       'W 78 St & Broadway', 'Riverside Blvd & W 67 St', 'E 48 St & 5 Ave',
       'Lexington Ave & E 63 St', 'Amsterdam Ave & W 82 St',
       'E 41 St & Madison Ave', 'E 40 St & Madison Ave',
       'Leonard St & Manhattan Ave', '21 St & 41 Ave'], dtype=object)

Using Googles api instead

In [7]:
missing_adds = []

In [8]:
for stn in samp.ix[samp.zipcode.isnull(), 'start station name'].unique():
    if stn in [a['stn'] for a in missing_adds]:
        continue
    print stn
    missing_adds.append({'stn': stn, 
            'address': goog.geocode(stn).address, 
             'lat_long':goog.geocode(stn)[1], 
             'zipcode': re.findall(r'\d{5}', goog.geocode(stn).address)[0]})


MacDougal St & Washington Sq
48 Ave & 5 St
Marcy Ave & MacDonough St
Central Park West & W 85 St
W 76 St & Columbus Ave
5 Ave & E 78 St
East End Ave & E 86 St
Huron St & Franklin St
W 78 St & Broadway
Riverside Blvd & W 67 St
E 48 St & 5 Ave
Lexington Ave & E 63 St
Amsterdam Ave & W 82 St
E 41 St & Madison Ave
E 40 St & Madison Ave
Leonard St & Manhattan Ave
21 St & 41 Ave


In [9]:
# Fill in the missing values in the dataframe
for dic in missing_adds:
    samp.set_value(samp['start station name']==dic['stn'], 'address', dic['address'])
    samp.set_value(samp['start station name']==dic['stn'], 'zipcode', dic['zipcode'])


In [10]:
# and convert to shapely geometry points (longitude, latitude)
samp['geo_points'] = samp.long_lat.map(geom.Point)

### Residential Vs Commercial Districts

Since nyc is divided into various types of zoning districts, I'm going to find which district each station lies in to see if it's residential or commercial.

Aquire zoning dimensions from: https://data.cityofnewyork.us/City-Government/Zoning-GIS-Data-Shapefile/kdig-pewd

In [11]:
zoning_data = gpd.read_file('Zoning GIS Data- Shapefile.geojson')
zoning_data.shape

(15714, 20)

In [12]:
# Now we want to map each bike location to a zone
print samp.shape
print zoning_data.shape

#overlay, zonedist

# this is going to map citibike station ids to zoning index
loc_zone = {}
for i in samp.index:
    if samp.ix[i, 'start station id'] in loc_zone:
        continue
    loc_zone[samp.ix[i, 'start station id']] = zoning_data[(zoning_data.geometry.contains(samp.ix[i, 'geo_points'])) & (~zoning_data.zonedist.isnull())].index
    


samp['zoning_idx'] = samp['start station id'].map(loc_zone)   

# change the type from an index object to an int
samp['zoning_idx'] = samp.zoning_idx.map(lambda i: list(i)[0])

(3184288, 20)
(15714, 20)


In [13]:
samp.head()

tripduration           starttime            stoptime  start station id  \
0          1346 2015-01-01 00:01:00 2015-01-01 00:24:00               455   
1           363 2015-01-01 00:02:00 2015-01-01 00:08:00               434   
2           346 2015-01-01 00:04:00 2015-01-01 00:10:00               491   
3           182 2015-01-01 00:04:00 2015-01-01 00:07:00               384   
4           969 2015-01-01 00:05:00 2015-01-01 00:21:00               474   

        start station name  start station latitude  start station longitude  \
0          1 Ave & E 44 St               40.750020               -73.969053   
1          9 Ave & W 18 St               40.743174               -74.003664   
2     E 24 St & Park Ave S               40.740964               -73.986022   
3  Fulton St & Waverly Ave               40.683178               -73.965964   
4          5 Ave & E 29 St               40.745168               -73.986831   

   end station id             end station name  end station latitude  \
0             265     Stanton St & Chrystie St             40.722293   
1             482              W 15 St & 7 Ave             40.739355   
2             505              6 Ave & W 33 St             40.749013   
3             399  Lafayette Ave & St James Pl             40.688515   
4             432            E 7 St & Avenue A             40.726218   

     ...      bikeid    usertype birth year  gender  \
0    ...       18660  Subscriber     1960.0       2   
1    ...       16085  Subscriber     1963.0       1   
2    ...       20845  Subscriber     1974.0       1   
3    ...       19610  Subscriber     1969.0       1   
4    ...       20197  Subscriber     1977.0       1   

                      lat_long                     long_lat  \
0  (40.75001986, -73.96905301)  (-73.96905301, 40.75001986)   
1  (40.74317449, -74.00366443)  (-74.00366443, 40.74317449)   
2  (40.74096374, -73.98602213)  (-73.98602213, 40.74096374)   
3   (40.68317813, -73.9659641)   (-73.9659641, 40.68317813)   
4   (40.7451677, -73.98683077)   (-73.98683077, 40.7451677)   

                                             address zipcode  \
0  Citi Bike - 1 Ave & E 44 St, 1st Avenue Tunnel...   10017   
1  Citi Bike - 9 Ave & W 18 St, 9th Avenue, Chels...   10011   
2  Citi Bike - E 24 St & Park Ave S, East 24th St...   10010   
3  884, Fulton Street, Adelphi, BK, Kings County,...   11238   
4  Citi Bike - 5 Ave & E 29 St, 5th Avenue, Korea...   10035   

                               geo_points zoning_idx  
0        POINT (-73.96905301 40.75001986)      14720  
1        POINT (-74.00366443 40.74317449)      10788  
2  POINT (-73.98602212999999 40.74096374)      10754  
3  POINT (-73.96596409999999 40.68317813)      14314  
4         POINT (-73.98683077 40.7451677)      14722  

[5 rows x 21 columns]

Combine the two so we now have a zone type for each bike station

The zone types are 
    - Residential
    - Commercial
    - Manufacturing
    - Park
    - Battery Park (bpc)

In [14]:
merged = pd.merge(samp, zoning_data[['zonedist']], left_on='zoning_idx', right_index = True)
merged['zone_type'] = merged.zonedist.map(lambda z: z[0] if z != 'PARK' else z)

# make sure zipcode is an integer
merged['zipcode'] = merged.zipcode.astype('int64')


with open('merged.pkl', 'wb') as f:
    pickle.dump(merged,f,2)

In [ ]:
with open('merged.pkl', 'rb') as f:
    merged = pickle.load(f)
    
merged.shape

### Population By Community District

Population is available on the Community District level from Nyc Open Data

In [15]:
# Geographical data for all the community districts, so we know where each district actually is.
cds = gpd.read_file("Community Districts.geojson")
cds.shape

(71, 4)

In [16]:
# Populations for each district
pops = pd.read_csv('New_York_City_Population_By_Community_Districts.csv')

boro_map = {'Manhattan':'1', 'Bronx':'2',  'Brooklyn':'3', 'Queens':'4', "Staten Island":'5'}
pops['CD_ID'] = pops.Borough.map(boro_map) + pops['CD Number'].astype(str).str.zfill(2)

In [17]:
# Add Community Districts to our dataframe so that we can merge with populations
stn_cd = {}

In [18]:
for i in merged.drop_duplicates('start station id').index:
    if merged.ix[i, 'start station id'] in stn_cd:
        continue
    stn_cd[merged.ix[i, 'start station id']] = cds[cds.geometry.contains(merged.ix[i, 'geo_points'])].boro_cd.values[0]

In [19]:
merged['CD'] = merged['start station id'].map(stn_cd)
merged.head()

tripduration           starttime            stoptime  start station id  \
0            1346 2015-01-01 00:01:00 2015-01-01 00:24:00               455   
369           133 2015-01-01 02:42:00 2015-01-01 02:44:00              2017   
447          1198 2015-01-01 03:15:00 2015-01-01 03:35:00              2017   
591           474 2015-01-01 06:53:00 2015-01-01 07:01:00              2017   
990           976 2015-01-01 10:45:00 2015-01-01 11:01:00              2017   

    start station name  start station latitude  start station longitude  \
0      1 Ave & E 44 St               40.750020               -73.969053   
369    E 43 St & 2 Ave               40.750224               -73.971214   
447    E 43 St & 2 Ave               40.750224               -73.971214   
591    E 43 St & 2 Ave               40.750224               -73.971214   
990    E 43 St & 2 Ave               40.750224               -73.971214   

     end station id            end station name  end station latitude ...   \
0               265    Stanton St & Chrystie St             40.722293 ...    
369             518             E 39 St & 2 Ave             40.747804 ...    
447             342  Columbia St & Rivington St             40.717400 ...    
591             164             E 47 St & 2 Ave             40.753231 ...    
990             402          Broadway & E 22 St             40.740343 ...    

     gender                     lat_long                     long_lat  \
0         2  (40.75001986, -73.96905301)  (-73.96905301, 40.75001986)   
369       1  (40.75022392, -73.97121414)  (-73.97121414, 40.75022392)   
447       2  (40.75022392, -73.97121414)  (-73.97121414, 40.75022392)   
591       1  (40.75022392, -73.97121414)  (-73.97121414, 40.75022392)   
990       1  (40.75022392, -73.97121414)  (-73.97121414, 40.75022392)   

                                               address  zipcode  \
0    Citi Bike - 1 Ave & E 44 St, 1st Avenue Tunnel...    10017   
369  Citi Bike - E 43 St & 2 Ave, East 43rd Street,...    10017   
447  Citi Bike - E 43 St & 2 Ave, East 43rd Street,...    10017   
591  Citi Bike - E 43 St & 2 Ave, East 43rd Street,...    10017   
990  Citi Bike - E 43 St & 2 Ave, East 43rd Street,...    10017   

                           geo_points zoning_idx zonedist  zone_type   CD  
0    POINT (-73.96905301 40.75001986)      14720     C5-2          C  106  
369  POINT (-73.97121414 40.75022392)      14720     C5-2          C  106  
447  POINT (-73.97121414 40.75022392)      14720     C5-2          C  106  
591  POINT (-73.97121414 40.75022392)      14720     C5-2          C  106  
990  POINT (-73.97121414 40.75022392)      14720     C5-2          C  106  

[5 rows x 24 columns]

District 164 represents central park, which does not have a given population statistic. Therefore i'm going to reasign the stations that fall within this district to their nearest community district with a given population

In [20]:
# For stations on CPW i'm going to assign the population data for District 107 
# which corresponds to the west side between 59th st and 110th
merged.loc[merged['start station name'].isin(['Central Park West & W 68 St',
                           'Central Park West & W 76 St', 'Central Park West & W 72 St']), 'CD'] = '107'


# The two stations on the east side are nearest to CD 108 which extends from ~ 59th - 86th st on the east side
merged.loc[merged['start station name'].isin(['5 Ave & E 63 St', '5 Ave & E 73 St']), 'CD'] = '108'


# CPS and 6th ave is nearest to district 105
merged.loc[merged['start station name'] == 'Central Park S & 6 Ave', 'CD'] = '105'

In [21]:
print merged.shape
print pops.shape

merged1 = pd.merge(merged, pops, left_on='CD', right_on = 'CD_ID')
print merged1.shape

(3184288, 24)
(59, 9)
(3178229, 33)


### Demographics by Zipcode

In [22]:
demographics = pd.read_csv('Zip_code_breakdowns.csv', usecols=['JURISDICTION NAME','COUNT PARTICIPANTS','PERCENT FEMALE', 'PERCENT MALE','PERCENT PACIFIC ISLANDER','PERCENT HISPANIC LATINO', 'PERCENT AMERICAN INDIAN',
'PERCENT ASIAN NON HISPANIC','PERCENT WHITE NON HISPANIC', 'PERCENT BLACK NON HISPANIC', 'PERCENT OTHER ETHNICITY'])
demographics.shape

(236, 11)

In [23]:
merged2 = pd.merge(merged1, demographics, left_on='zipcode', right_on='JURISDICTION NAME', how = 'left')
merged2.shape

(3178229, 44)

In [24]:
merged2.isnull().sum()

tripduration                       0
starttime                          0
stoptime                           0
start station id                   0
start station name                 0
start station latitude             0
start station longitude            0
end station id                     0
end station name                   0
end station latitude               0
end station longitude              0
bikeid                             0
usertype                           0
birth year                    364383
gender                             0
lat_long                           0
long_lat                           0
address                            0
zipcode                            0
geo_points                         0
zoning_idx                         0
zonedist                           0
zone_type                          0
CD                                 0
Borough                            0
CD Number                          0
CD Name                            0
1

Lots of missing values.

Additionally, some of the demographic surveys have 0 participants and hence all the values are therefore 0. We will treat those as missing values

In [25]:
# FILL NAs 
merged2.fillna({'COUNT PARTICIPANTS': 0}, inplace = True)

# Fill with mean for our features of interest
null_cols = ['PERCENT FEMALE','PERCENT MALE','PERCENT PACIFIC ISLANDER','PERCENT HISPANIC LATINO',
'PERCENT AMERICAN INDIAN', 'PERCENT ASIAN NON HISPANIC','PERCENT WHITE NON HISPANIC','PERCENT BLACK NON HISPANIC',
'PERCENT OTHER ETHNICITY']

merged2.loc[:,null_cols] = merged2[null_cols].fillna(merged2[null_cols].mean())

# Rows where the number of participants is 0. Replace them with the mean as well
merged2.loc[merged2['COUNT PARTICIPANTS'] == 0, null_cols] =  merged2.ix[merged2['COUNT PARTICIPANTS'] == 0, null_cols].apply(lambda c: c.replace(0,np.nan)).fillna(merged2[null_cols].mean())


Parse the datetime feature into season, day of the week, and date (without time)

In [26]:
s = zip(range(1,13),['winter','winter']+['spring']*3 + ['summer']*3 + ['fall'] *3 + ['winter'])
seasons = dict(s)


dow = dict(zip(range(7), ['monday','tuesday','wednesday','thursday','friday','saturday','sunday']))

merged2['season'] = merged2.starttime.dt.month.map(seasons)
merged2['dow'] = merged2.starttime.dt.dayofweek.map(dow)
merged2['date'] = merged2.starttime.dt.date

### Weather Data

Collected historical weather data from https://www.wunderground.com/

In [27]:
weather = []
with open('get_weather_data/historicalWeather.jl', 'r') as f:
    for line in f:
        weather.append(json.loads(line))
        
# hmmm we have 7 extra days...      
len(weather)

372

In [28]:
weather = pd.DataFrame(weather)
# change data to numbers
weather.iloc[:,1:] = weather.iloc[:,1:].apply(lambda c: pd.to_numeric(c, errors='coerce'))


def reorder_date(d):
    y,m,d = d.split('/')
    return '/'.join([m,d,y])
# parse dates
weather['date'] = pd.to_datetime(weather.date.map(reorder_date), errors = 'coerce')

print weather.isnull().sum()


date             7
dew_point        0
max_temp         0
mean_temp        0
min_temp         0
month_precip    11
month_snow       7
precip          22
snow            14
wind_speed       2
dtype: int64


In [29]:
weather = weather[[u'date', u'dew_point', u'max_temp',
 u'mean_temp', u'min_temp',u'precip', u'month_precip',
 u'snow', u'month_snow', u'wind_speed']]

In [30]:
# drop not real dates eg 2/30
weather.drop(weather[weather.date.isnull()].index, inplace = True)
weather.reset_index(inplace=True)
weather.shape # HOORAY

(365, 11)

#### Fill NAs

In [31]:
# all the month precip nulls are in one week in may when it did not rain. So filling with 0s
weather.fillna({'month_precip':0}, inplace = True)

# Month Snow nulls are in first week of January when it did not snow and end of December where there was no snow
# so filling with zeros
weather.fillna({'month_snow':0, 'snow':0}, inplace = True)

# Filling wind speed with the median value
print weather.wind_speed.median()
weather.fillna({'wind_speed': weather.wind_speed.median()}, inplace = True)
weather.isnull().sum()


#checked by hand, all the precip nulls are 0
weather.fillna({'precip':0}, inplace = True)

weather.isnull().sum()

5.0


index           0
date            0
dew_point       0
max_temp        0
mean_temp       0
min_temp        0
precip          0
month_precip    0
snow            0
month_snow      0
wind_speed      0
dtype: int64

Join the Weather Data to the rest

In [43]:
print weather.shape
print merged2.shape
merged3 = pd.merge(weather, merged2, left_on = weather.date.dt.date, right_on = 'date')
merged3.shape

(365, 11)
(3178229, 47)


(3178229, 58)

In [44]:
merged3.drop(['index','date_x'], inplace = True, axis = 1)
merged3.rename(columns={"date_y": 'date'}, inplace=True)

Lastly I want to make sure that each station is represented by it's usage. Creating 450+ dummy variables for each of the stations would be too much, so instead I'm going to just create 2 new columns. One that represents the mean daily usage of the given station, the other just an indicator of whether or not that station is one of the top 6 most used stations.

One thing to consider is how to calculate those 2 columns. Here I'm using the same data to calculate the mean usage and largest stations - but in practice I'd want to use the data from the same month the year prior.

In [51]:
usg = merged3.groupby(['date', 'start station id']).size().reset_index()
usg.columns = ['date','station','usg']
station_usg = usg.groupby('station').usg.mean()
station_usg.head()

station
72      74.073171
79      61.090164
82      26.577236
83      29.577236
116    136.545455
Name: usg, dtype: float64

In [52]:
top6 = station_usg.sort_values(ascending=False)[:6].index.tolist()
top6 = {s: top6.index(s) for s in top6}
top6

{293: 3, 435: 4, 497: 5, 519: 2, 521: 1, 3230: 0}

In [53]:
merged3['station_mean_usg'] = merged3['start station id'].map(station_usg)
merged3['busy_stn'] = 0
merged3.loc[merged3['start station id'].isin(top6), 'busy_stn'] = 1

In [54]:
merged3[['dew_point','max_temp','mean_temp','min_temp']] = merged3[['dew_point','max_temp','mean_temp','min_temp']].astype('int64')

In the next notebook, I will build a model to predict how many riders will use each station on a given day. This will be a good tool to estimate bike demand at each station based on the factors we have gathered thus far.

In [55]:
with open('data.pkl', 'wb') as f:
    pickle.dump(merged3, f, 2)